### PR TITLE
[WIP] ESNext with `async` syntax. (Works on #47)

### PR DESCRIPTION
**Pull request by jayands**
Whew! This is a big PR. It's broken, though, in that it incorrectly resolves an indirect circular dependency: Document -> Block ->Template -> Document. (I think that's what the cycle was, at least.) That said, there are a lot of things that work individually for this code, as I copied it over from a TypeScript test repo I created locally. This means that, with very minimal retooling it can be converted to TS. (seriously; only a `ts{config,lint}.json` file and the respective dependencies, the renaming of the files in `src`, and any references of type `mixed` being changed (back) to type `object`, and it should be good to go, although there are still errors)

Still, the *static* errors should be minimal, barring VSCode's complaining about using types in non-TypeScript files.

### CODE REVIEW

File: src/liquid/block.js [hunk 1]
```diff
@@ -0,0 +1,218 @@
+// @flow
+
+import { TagEnd, TagStart, VariableEnd, VariableStart } from './regexps';
+
+import Context from './context';
+import { SyntaxError } from './errors';
+import Tag from './tag';
+import Template from './template';
```
Comment (sirlantis) [inline src/liquid/block.js]: This is the circular dependency?

It's not used as a value further below, only as a type annotation. In TypeScript this `import` would have been removed in that case, `flow` probably doesn't do that.

File: src/liquid/block.js [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+      }
+      return Promise.resolve()
+      .then(() => token.render(context))
+      .then(s => accumulator.push(s), e => accumulator.push(context.handleError(e)));
+    }).then(() => accumulator);
+  }
+*/
+  async renderAll(list: Promise<Tag>[], context: Context): Promise<string[]> {
+    const accumulator: string[] = [];
+    return promiseEach(list, async (token: Tag | string) => {
```
Comment (sirlantis) [inline src/liquid/block.js]: `promiseEach` is probably unnecessary when you have await.

```
for (const promise of list) {
  const token = await promise;
  // ...
}
```

File: src/liquid/block.coffee [hunk 3]
```diff
@@ -1,6 +1,6 @@
 Liquid = require("../liquid")
```
Comment (sirlantis) [inline src/liquid/block.coffee]: I guess you kept this file because the `block.js` was not done yet?

File: src/liquid/engine.js [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+import BlankFileSystem from './blank_file_system';
+import Block from './block';
+import StandardFilters from './standard_filters';
+import Tag from './tag';
+import Tags from './tags';
+import Template from './template';
+
+class Engine {
+  tags: Map<string, Tag> = new Map();
+  Strainer: Map<string, Function> = new Map();
```
Comment (sirlantis) [inline src/liquid/engine.js]: This was supposed to be a class before. I think we could change this now to be called `filters` instead.

File: src/liquid/tag.js [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+    if (this.afterParse) {
+      await this.afterParse(...args);
+    }
+  }
+  /* eslint-disable class-methods-use-this */
+  async afterParse() { /* no-op */ }
+  async beforeParse() { /* no-op */ }
+  async parse() { return false; }
+  /* eslint-disable no-unused-vars */
+  async render(context: Context) {
```
Comment (sirlantis) [inline src/liquid/tag.js]: Can you use `_context` to make `eslint` not complain?

File: src/liquid/tag.js [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+  async parseWithCallbacks(...args: mixed[]) {
+    if (this.beforeParse) {
+      await this.beforeParse(...args);
+    }
+    this.parse(...args);
+    if (this.afterParse) {
+      await this.afterParse(...args);
+    }
+  }
+  /* eslint-disable class-methods-use-this */
```
Comment (sirlantis) [inline src/liquid/tag.js]: Maybe we should disable that rule globally.

File: src/liquid/condition.js [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+class Condition {
+
+  static LITERALS: Map<string, ((v: any) => boolean)> = new Map([
+    ['empty', (v: any) => {
+      if (v == null) return true;
+      return !(v.length > 0);
+    }],
+  ['blank', (v: any) => !v || v.toString().length === 0],
+  ]);
+  // tslint:disable-next-line:max-line-length
```
Comment (sirlantis) [inline src/liquid/condition.js]: Probably not needed anymore?

File: src/liquid/condition.js [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+  static OPERATORS: Map<string, ((left: any, right: any) => boolean)> = new Map([
+  ['==', (left: any, right: any) => Condition.equalVariables(left, right)],
+  ['is', (left: any, right: any) => Condition.equalVariables(left, right)],
+  ['!=', (left: any, right: any) => !Condition.equalVariables(left, right)],
+  ['<>', (left: any, right: any) => !Condition.equalVariables(left, right)],
+  ['isnt', (left: any, right: any) => !Condition.equalVariables(left, right)],
+  ['<', (left: any, right: any) => left < right],
+  ['>', (left: any, right: any) => left > right],
+  ['<=', (left: any, right: any) => left <= right],
+  ['>=', (left: any, right: any) => left >= right],
```
Comment (sirlantis) [inline src/liquid/condition.js]: lines above should probably be indented by one level

File: src/liquid/condition.js [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+
+  async interpretCondition(left: string, right: string, op: string, context: Context) {
+    if (op == null) {
+      return !!(await this.resolveVariable(left, context));
+    }
+    const operation = Condition.OPERATORS.get(op);
+    if (operation == null) {
+      throw new Error(`Unknown operator ${op}`);
+    }
+    const lhs = await this.resolveVariable(left, context);
```
Comment (sirlantis) [inline src/liquid/condition.js]: `Promise.all` would allow for parallel execution of those resolves.

File: src/liquid/context.js [hunk 10]
```diff
@@ -0,0 +1,319 @@
...
+import { ContextError, Error, FilterNotFound, SyntaxError } from './errors';
+
+import Drop from './drop';
+import Engine from './engine';
+import Range from './range';
+import { VariableParser } from './regexps';
+
+class Context {
+  static LITERALS = new Map([
+  ['nil', null],
```
Comment (sirlantis) [inline src/liquid/context.js]: This and following lines: Indent one level?

File: src/liquid/context.js [hunk 11]
```diff
@@ -0,0 +1,319 @@
...
+  /**
+  * Adds filters to this context.
+  *
+  * Note that this does not register the filters with the main Template object.
+  * See {@link Template#registerFilter} for that.
+  */
+  registerFilters(filters: Map<string, Function>) {
+    const self = this;
+    Object.entries(filters)
+    .map(([k, v]) => self.strainer.set(k, v));
```
Comment (sirlantis) [inline src/liquid/context.js]: Indent one level?

File: src/liquid/context.js [hunk 12]
```diff
@@ -0,0 +1,319 @@
...
+    Object.entries(newScope).forEach(([key, val]) => currentScope.set(key, val));
+    this.scopes[0] = currentScope;
+  }
+  pop() {
+    if (this.scopes.length <= 1) {
+      throw new ContextError();
+    }
+    return this.scopes.shift();
+  }
+  lastScope() {
```
Comment (sirlantis) [inline src/liquid/context.js]: This could be a get lastScope now.

File: src/liquid/tags/for.js [hunk 13]
```diff
@@ -0,0 +1,129 @@
...
+      scan(markup, TagAttributes).forEach((attr) => {
+        self.attributes[attr[0]] = attr[1];
+      });
+    } else {
+      throw new SyntaxError(SyntaxHelp);
+    }
+    this.forBlock = [];
+    this.nodelist = this.forBlock;
+  }
+  render(context) {
```
Comment (sirlantis) [inline src/liquid/tags/for.js]: This function could probably benefit a lot from `async`

File: src/liquid/context.js [hunk 14]
```diff
@@ -0,0 +1,319 @@
...
+    return this.scopes[this.scopes.length - 1];
+  }
+  stack(f: Function, newScope: Map<string, any> = new Map()) {
+    let popLater = false;
+    try {
+      const self = this;
+      this.push(newScope);
+      const result = f();
+      if (result != null && result.nodeify != null) {
```
Comment (sirlantis) [inline src/liquid/context.js]: This was a check whether result is a promise. I guess `nodeify` does not work anymore.

File: src/liquid/context.js [hunk 15]
```diff
@@ -0,0 +1,319 @@
...
+  pop() {
+    if (this.scopes.length <= 1) {
+      throw new ContextError();
+    }
+    return this.scopes.shift();
+  }
+  lastScope() {
+    return this.scopes[this.scopes.length - 1];
+  }
+  stack(f: Function, newScope: Map<string, any> = new Map()) {
```
Comment (sirlantis) [inline src/liquid/context.js]: Making stack `async` might allow the code here to become easier.

You might need to change type of `f` to see if anyone doesn't return a promise and then make them so.

```
this.push(newScope);

try {
  return await f();
} finally {
  this.pop();
}
```

File: src/liquid/context.js [hunk 16]
```diff
@@ -0,0 +1,319 @@
...
+      // The iterator walks through the parsed path step
+      // by step and waits for promises to be fulfilled.
+    async function iterator(object: any, index: number): Promise<any> {
+      if (index < parts.length) {
+        const tmp: any = await mapper(parts[index], object);
+        return iterator(tmp, index + 1);
+      }
+      return object;
+    }
+    try {
```
Comment (sirlantis) [inline src/liquid/context.js]: The `iterator` can be removed now with `await`.

```
for (const part of parts) {
  object = await mapper(part, object);
}

return object;
```

File: src/liquid/tags/capture.js [hunk 17]
```diff
@@ -0,0 +1,32 @@
...
+    const match = SYNTAX.exec(markup);
+    if (match) {
+      this.to = match[1];
+    } else {
+      throw new SyntaxError(SYNTAX_HELP);
+    }
+  }
+  render(context: Context) {
+    // const chunks = await super.render(context);
+    // const output = Helpers.toFlatString(chunks);
```
Comment (sirlantis) [inline src/liquid/tags/capture.js]: Why are these lines commented out?

File: src/liquid/context.js (hunk 16 above)
Comment (jayands) [inline src/liquid/context.js]: Could that be further simplified as a reducer function?
```js
return parts.reduce(async part => await mapper(part, object));
```
Or do you think that's  incorrect, too ambiguous and/or too difficult to reason about? It seemed to work in the REPL with simple items.

File: src/liquid/context.js (hunk 15 above)
Comment (jayands) [inline src/liquid/context.js]: You're correct. Implemented.

File: src/liquid/context.js (hunk 12 above)
Comment (jayands) [inline src/liquid/context.js]: You're correct. Implemented. Also, I think it's only called in one spot, so is it supposed to be exposed?

File: src/liquid/block.coffee (hunk 3 above)
Comment (jayands) [inline src/liquid/block.coffee]: You are correct on that.

File: src/liquid/block.js (hunk 1 above)
Comment (jayands) [inline src/liquid/block.js]: Funnily enough, that's *exactly* what happens in TypeScript---I've actually gotten it down to two errors in TypeScript, only one of which is actually related to the actual code being worked on:

```log
src/liquid/standard_filters.ts(1,20): error TS7016: Could not find a declaration file for module 'moment-strftime'. '/home/jayands/Projects/typescript/ts.liquid-node/node_modules/moment-strftime/lib/moment-strftime.js' implicitly has an 'any' type.
  Try `npm install @types/moment-strftime` if it exists or add a new declaration (.d.ts) file containing `declare module 'moment-strftime';`
src/liquid/tags/case.ts(70,22): error TS2339: Property 'forEach' does not exist on type 'string'.
```

File: src/liquid/context.js (hunk 16 above)
Comment (sirlantis) [inline src/liquid/context.js]: `reduce` is certainly the thing we do here, but it's not promise aware. I think we need to tweak it a bit to work correctly:

```
return parts.reduce(async (acc, part) => mapper(part, await acc), Promise.resolve(object))
```

Now `acc` is always of type `Promise<?>` and will expose the behavor from the `for` loop before.

File: src/liquid/context.js (hunk 12 above)
Comment (sirlantis) [inline src/liquid/context.js]: Capture and Assign use them as well.

File: src/liquid/block.js (hunk 1 above)
Comment (sirlantis) [inline src/liquid/block.js]: @jayands I'll try to get a closer look at it this weekend.

**Comment by sirlantis** [comment]
Thanks for the updates @jayands. Sorry for the silence again, have been traveling and didn't find the time to keep up with GH. I'll try to take a closer look on it this weekend again. It's looking really good from what I can tell.

**Comment by zeke** [comment]
Hi @jayands. Thanks for the PR. We recently migrated the codebase from CoffeeScript to JavaScript, see #71.  Because of that change, this PR will no longer be mergable as-is If this change is still something you'd like to see included in the project, please comment here to restart the conversation. Thanks!

**Pull request closed without merging**